crates/sdk/src/signing.rs

```diff
@@ -702,61 +701,101 @@ fn format_outputs(output: &mut Vec<String>) {
     }
 }
 
-enum TokenTransfer<'a> {
-    Transparent(&'a token::TransparentTransfer),
-    Shielded,
-    Shielding(&'a token::ShieldingTransfer),
-    Unshielding(&'a token::UnshieldingTransfer),
+enum TransferSide<'a> {
+    Source(&'a Address),
+    Target(&'a Address),
 }
 
+struct TokenTransfer<'a>(&'a token::Transfer);
+
 impl TokenTransfer<'_> {
-    fn source(&self) -> Option<&Address> {
-        match self {
-            TokenTransfer::Transparent(transfer) => Some(&transfer.source),
-            TokenTransfer::Shielded => None,
-            TokenTransfer::Shielding(transfer) => Some(&transfer.source),
-            TokenTransfer::Unshielding(_) => None,
-        }
+    fn sources(&self) -> Vec<&Address> {
```

Comment:
Question here: should we ensure that in the vectors return by `sources` and `targets` there are no duplicates? Probably not because of the hardware wallet?

Reply:
> Question here: should we ensure that in the vectors return by `sources` and `targets` there are no duplicates? Probably not because of the hardware wallet?

I think you're right. We should ensure there are no duplicates in `sources` nor `targets`. Otherwise `make_ledger_token_transfer_endpoints` might print the same source or target repeatedly. Would you be able to look at #3459 ? I think this issue is inadvertently fixed there...